File: Mandatory/013 - Balanced Brackets/BalancedBrackets.rb
class Node
  attr_accessor :value, :next      

  def initialize(value)
    @value = value
    @next = nil
  end
end

class LinkedList
  attr_accessor :first, :last, :count

  def initialize
    @first = nil
    @last = nil
    @count = 0
  end

  def add(value)
    if @last
      new_node = Node.new(value)
      @last.next = new_node
      @last = @last.next      
    else
      @first = Node.new(value)
      @last = @first
    end
    @count += 1
  end

  def add_at(index, value)
    if @last
      if index == 0
        new_node = Node.new(value)
        new_node.next = @first      
        @first = new_node
        @count += 1
      elsif index == @count
        new_node = Node.new(value)
        @last.next = new_node
        @last = new_node
        @count += 1
      elsif index > 0 && index < @count
        new_node = Node.new(value)
        previous_node = get_node(index)
        new_node.next = previous_node.next
        previous_node.next = new_node
        @count += 1
      end
    elsif index == 0
      @first = Node.new(value)
      @last = @first
      @count += 1
    end    
  end

  def remove(index)
    if @last
      if index == 0
        old_node = @first
        @first = @first.next
        old_node = nil
        @count -= 1
      elsif index == @count - 1
        old_node = @last
        @last = get_node(index)
        old_node = nil
        @count -= 1
      elsif index > 0 && index < @count
        previous_node = get_node(index)
        old_node = previous_node.next
        previous_node.next = old_node.next
        old_node = nil
        @count -= 1
      end    
    end

    if @count == 0
      @first = nil
      @last = nil
    elsif @count == 1
      @first = @last if @first
      @last = @first if @last
    end
  end

  def get(index)
    node = get_node(index)
    return node.value if node
    nil
  end

  private

  def get_node(index)
    if @first
      if index >= 0 && index < @count      
        current_node = @first
        index.times { |item| current_node = current_node.next }
        current_node
      else
        nil
      end
    else
      nil
    end
  end
end

class Stack
  attr_accessor :list

  def initialize
    @list = LinkedList.new
  end

  def push(value)       
    if @list.first
      @list.add_at(0, value)
    else
      @list.add(value)
    end    
  end

  def pop
    popped = -1
    if @list.first
      popped = @list.get(0)      
      @list.remove(0)      
    end
    popped
  end
end

def balanced_brackets?(evalString)  
  array_string = evalString.split('')  
  evaluated_array = []
  
  array_string.each do |st| 
    if ['(', ')', '{', '}', '[', ']'].any?(st)
      evaluated_array.push(st)
    end
  end
  
  forced_finish = false
  temp_stack = Stack.new

  evaluated_array.each do |item| 
    if item == '('
      temp_stack.push(')')
    elsif item == '['
      temp_stack.push(']')
    elsif item == '{'
      temp_stack.push('}')
    else
      if temp_stack.list.first.nil?
        forced_finish = true
        break
      elsif item == temp_stack.list.first.value
        temp_stack.pop
      end
    end    
  end
  
  temp_stack.list.first.nil? && !forced_finish
end


puts balanced_brackets?('(hello)[world]')
# => true

puts balanced_brackets?('([)]')
# => false

puts balanced_brackets?('[({}{}{})([])]')
# => true

File: Mandatory/013 - Balanced Brackets/BalancedBrackets.py
class Node:
  def __init__(self, value):
    self.value = value
    self.next = None
  
class LinkedList:
  def __init__(self):
    self.first = None
    self.last = None
    self.count = 0


  def add(self, value):
    if (self.last):
      new_node = Node(value)
      self.last.next = new_node
      self.last = self.last.next
    else:
      self.first = Node(value)
      self.last = self.first

    self.count += 1


  def add_at(self, index, value):
    if (self.last):
      if (index == 0):
        new_node = Node(value)
        new_node.next = self.first
        self.first = new_node
        self.count += 1
      elif (index == self.count):
        new_node = Node(value)
        self.last.next = new_node
        self.last = new_node
        self.count += 1
      elif (index > 0 and index < self.count):
        new_node = Node(value)
        previous_node = self.__get_node(index)
        new_node.next = previous_node.next
        previous_node.next = new_node
        self.count += 1  
    elif (index == 0):
      self.first = Node(value)
      self.last = self.first
      self.count += 1


  def remove(self, index):
    if (self.last):
      if (index == 0):
        old_node = self.first
        self.first = self.first.next
        old_node = None
        self.count -= 1
      elif (index == self.count - 1):
        old_node = self.last
        self.last = self.__get_node(index)
        old_node = None
        self.count -= 1
      elif (index > 0 and index < self.count):
        previous_node = self.__get_node(index)
        old_node = previous_node.next
        previous_node.next = old_node.next
        old_node = None
        self.count -= 1

    if (self.count == 0):
      self.first = None
      self.last = None
    elif (self.count == 1):
      if (self.first): self.first = self.last
      if (self.last): self.last = self.first


  def get(self, index):
    node = self.__get_node(index)
    if (node): 
      return node.value
    else:
      return None

  def __get_node(self, index):
    if (self.first):
      if (index >= 0 and index < self.count):
        current_node = self.first
        for _ in range(0, index):
          current_node = current_node.next
        
        return current_node
      else:
        return None
    else:
      return None

class Stack:
  def __init__(self):
    self.list = LinkedList()

  def push(self, value):
    if (self.list.first):
      self.list.add_at(0, value)
    else:
      self.list.add(value)

  def pop(self):
    popped = -1
    if (self.list.first):
      popped = self.list.get(0)
      self.list.remove(0)
    
    return popped


def balanced_brackets(evalString):
  array_string = list(evalString)
  evaluated_array = []
  forced_finish = False
  temp_stack = Stack()

  for st in array_string:
    if (st in ['(', ')', '{', '}', '[', ']']):
      evaluated_array.append(st)

  for item in evaluated_array:  
    if (item == '('):
      temp_stack.push(')')
    elif (item == '['):
      temp_stack.push(']')
    elif (item == '{'):
      temp_stack.push('}')
    else:
      if (not temp_stack.list.first):
        forced_finish = true
        break
      elif (item == temp_stack.list.first.value):
        temp_stack.pop()
  
  return ((not temp_stack.list.first) and (not forced_finish))


print(balanced_brackets('(hello)[world]'))
# => true

print(balanced_brackets('([)]'))
# => false

print(balanced_brackets("[({}{}{})([])]"))
# => true  
